nefertari/elasticsearch.py

```diff
@@ -248,91 +237,178 @@ def get_document_proxies_by_type(cls, doc_type):
-        transaction.addAfterCommitHook(self.get_hook(), kws={'transaction': transaction})
-        self.registry[transaction] = [es_action]
+            for item in op_type_actions:
+                types.add(item.entity_type)
+            for type_ in types:
+                results = [item for item in ESActionRegistry.get_latest_change(filter(lambda i: i.entity_type == type_,
+                                                                                      op_type_actions))]
+                es_data.extend(results)
+
+        es_data = ESActionRegistry.prepare_for_deletion(es_data)
+        flat_actions = [data.action for data in es_data]
+
+        if ES.settings.get('enable_refresh_query'):
+            refresh_enabled = ES.settings.asbool('enable_refresh_query')
+        else:
+            refresh_enabled = False
+
+        kwargs = {
+            'client': ES.api,
+            'actions': flat_actions,
+
+        }
+
+        if request:
+            query_params = request.params.mixed()
+            refresh_index = query_params.get('_refresh_index', True)
+        else:
+            refresh_index = True
+
+        if refresh_enabled:
+            kwargs['refresh'] = refresh_index
+        helpers.bulk(**kwargs)
+
+        log.debug('Successfully executed {} elasticsearch actions'.format(list(map(lambda x: (x['_op_type'], x['_type'], x['_id']),kwargs['actions']))))
 
     @staticmethod
-    def force_indexation(actions):
-        failed_actions = []
+    def get_latest_change(data):
+        ids = {}
+        for item in data:
+            if item.id in ids:
+                ids[item.id].append(item)
+            else:
+                ids[item.id] = [item]
 
-        for action in actions:
-            successful, error = action()
+        for i in ids:
+            yield sorted(ids[i], key=lambda x: x.created_at).pop()
+
+    @staticmethod
+    def prepare_for_deletion(es_data):
+        to_delete = list(filter(lambda i: i.op_type == 'delete', es_data))
+        results = []
+
+        if not to_delete:
+            return es_data
+
+        def should_be_deleted(data):
+            for item in to_delete:
+                if data.entity_type == item.entity_type and data.id == item.id and data.op_type != 'delete':
+                    return True
+            return False
 
-            if not successful:
-                # handle failed action, maybe schedule reindex round
-                failed_actions.append((action, error))
-            if failed_actions:
-                raise ESBulkException([error for action, error in failed_actions])
+        for item in reversed(es_data):
+            if not should_be_deleted(item):
+                results.append(item)
 
+        return results
 
-class ESBulkException(ElasticsearchException):
-    def __init__(self, errors):
-        self.errors = errors
+
+class ESData:
+
+    def __init__(self, action: dict, created_at):
+        self.action = action
+        self.created_at = created_at
```

Comment:
What do you think about renaming `created_at` for `creation_time`? I think `created_at` can be confusing since we have this property in our entities